pandas/core/dtypes/dtypes.py

```diff
@@ -817,6 +817,63 @@ def __eq__(self, other: Any) -> bool:
             and tz_compare(self.tz, other.tz)
         )
 
+    def __from_arrow__(
+        self, array: pyarrow.Array | pyarrow.ChunkedArray
+    ) -> DatetimeArray:
+        """
+        Construct DatetimeArray from pyarrow Array/ChunkedArray.
+
+        Note: Values corresponding to the integer representation of ``NaT``
+        (e.g. one nanosecond before :attr:`pandas.Timestamp.min`) are converted
+        to ``NaT``, regardless of the null indicator in the pyarrow array.
+
+        Parameters
+        ----------
+        array : pyarrow.Array or pyarrow.ChunkedArray
+            The Arrow array to convert to DatetimeArray.
+
+        Returns
+        -------
+        extension array : DatetimeArray
+        """
+        import pyarrow
+
+        from pandas.core.arrays import DatetimeArray
+        from pandas.core.arrays.arrow._arrow_utils import (
+            pyarrow_array_to_numpy_and_mask,
+        )
+
+        pa_type = array.type
+        pa_unit = "ns"
+        if pyarrow.types.is_timestamp(pa_type):
+            pa_unit = pa_type.unit
+
+        if isinstance(array, pyarrow.Array):
+            chunks = [array]
+        else:
+            chunks = array.chunks
+
+        np_dtype = np.dtype(f"datetime64[{self._unit}]")
+        results = []
+        for arr in chunks:
+            if arr.nbytes == 0:
+                continue
+            data, mask = pyarrow_array_to_numpy_and_mask(
+                arr, dtype=np.dtype(f"datetime64[{pa_unit}]")
+            )
```

Comment:
I think we can rely on the built-in conversion from pyarrow, and call `arr.to_numpy(zero_copy_only=False)`. That should already handle the inclusion of NaT (and more efficiently), and also work for chunked arrays (so no need to iterate manually)

The only thing you then afterwards need to handle here is the tz_localize

Reply:
Thanks! That really simplifies things.